shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -555,89 +555,92 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
-        val typeRefDef = ctx.tyDefs.getOrElse(defn.name, throw new Exception(s"Cannot find $defn type definition in context"))
-        targs.zip(typeRefDef.tvars).foreach { case (targ, tvar) =>
-          val tvarVariance = if (curVariance === VarianceInfo.contra())
-              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.contra())
-            else
-              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.co())
-          updateVariance(targ, tvarVariance)
-        }
-      case ComposedType(pol, lhs, rhs) =>
-        updateVariance(lhs, curVariance)
-        updateVariance(rhs, curVariance)
-      case TypeBounds(lb, ub) =>
-        updateVariance(lb, curVariance.flip(VarianceInfo.contra()))
-        updateVariance(ub, curVariance.flip(VarianceInfo.co()))
-      case ArrayType(inner) => fieldVarianceHelper(inner, curVariance)
-      case TupleType(fields) => fields.foreach {
-          case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit =
+      trace(s"updateVariance(ty: $ty of ${ty.getClass()}, curVariance: $curVariance)(${tyDef.nme.name}.${tyDef.tvarVariances})") {
+        ty match {
+          case ProxyType(underlying) => updateVariance(underlying, curVariance)
+          case TraitTag(_) | ClassTag(_, _) => ()
+          case ExtrType(pol) => ()
+          case t: TypeVariable => val _ = tyDef.tvarVariances.updateWith(t) {
+            case None => Some(curVariance)
+            case Some(value) =>
+              val newVariance = value && curVariance
+              if (newVariance =/= curVariance) varianceUpdated = true;
+              Some(newVariance)
+          }
+          case RecordType(fields) => fields.foreach {
+            case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
+          }
+          case NegType(negated) =>
+            updateVariance(negated, curVariance.flip)
+          case TypeRef(defn, targs) =>
+            val typeRefDef = ctx.tyDefs.getOrElse(defn.name, throw new Exception(s"Cannot find $defn type definition in context"))
+            targs.zip(typeRefDef.tvars).foreach { case (targ, tvar) =>
+              typeRefDef.tvarVariances.getOrElse(tvar, VarianceInfo.bi) match {
+                case in @ VarianceInfo(false, false) => updateVariance(targ, in)
+                case VarianceInfo(true, false) => updateVariance(targ, curVariance)
+                case VarianceInfo(false, true) => updateVariance(targ, curVariance.flip)
+                case VarianceInfo(true, true) => ()
+              }
+            }
+          case ComposedType(pol, lhs, rhs) =>
+            updateVariance(lhs, curVariance)
+            updateVariance(rhs, curVariance)
+          case TypeBounds(lb, ub) =>
+            updateVariance(lb, curVariance.flip)
+            updateVariance(ub, curVariance)
```

Comment:
So the funny thing with `TypeBounds` is that the LB and UB are always at the same polarities:
```suggestion
            updateVariance(lb, VarianceInfo.contra)
            updateVariance(ub, VarianceInfo.co)
```